src/lib/staged_ledger/staged_ledger.ml

```diff
@@ -1238,11 +1319,47 @@ module T = struct
             ~get_batch:(Ledger.get_batch ledger) )
       |> Deferred.return
     in
-    let%map xs = Verifier.verify_commands verifier cs in
+    (* PERF:
+         We filter on the txns to verifier and clear out already verified cases,
+         hence we waste less time interacting with proof layers.
+         More info check https://www.notion.so/o1labs/Verification-of-zkapp-proofs-prior-to-block-creation-196e79b1f910807aa8aef723c135375a *)
+    let last_index = List.length cs - 1 in
```

Comment:
I suggest to extract this functionality to a separate function.

That function would have signature like:

```ocaml
(**
 * [process_separately] splits the list in two, and applies transformations to both parts, then it merges the list back in the same order it was originally.
 * [process_true] and [process_false] are expected to return the same number of elements processed in the same order.
 *)
val process_separately : separating_predicate:('a -> bool) -> process_true:(`a list -> `b list) -> process_false:(`a list -> `b list) -> 'b list
```

Let's use the following function for splitting:
```ocaml
val partition : 'a list -> f:('a -> bool) -> 'a list * 'a list
```

With it before splitting we'd need to `List.zip` with a series of indices (`List.init n ~f:ident`).

Reply:
After performing the extraction, it's extremely easy to cover this function with unit test(s) based on numbers.